server/src/internalClusterTest/java/org/elasticsearch/monitor/metrics/NodeIndexingMetricsIT.java

```diff
@@ -77,6 +90,197 @@ protected Settings nodeSettings(int nodeOrdinal, Settings otherSettings) {
+                assertThat(e.status(), is(RestStatus.CONFLICT));
+            }
+        }
+
+        // simulate async apm `polling` call for metrics
+        plugin.collect();
+
+        // there are no indexing (version conflict) failures reported because only gets/updates/deletes generated the conflicts
+        // and those are not "indexing" operations
+        var indexingFailedTotal = getSingleRecordedMetric(plugin::getLongAsyncCounterMeasurement, "es.indexing.indexing.failed.total");
+        assertThat(indexingFailedTotal.getLong(), equalTo(0L));
+        var indexingFailedDueToVersionConflictTotal = getSingleRecordedMetric(
+            plugin::getLongAsyncCounterMeasurement,
+            "es.indexing.indexing.failed.version_conflict.total"
+        );
+        assertThat(indexingFailedDueToVersionConflictTotal.getLong(), equalTo(0L));
+    }
+
+    public void testMetricsForIndexingVersionConflicts() {
+        final String dataNode = internalCluster().startNode();
+        ensureStableCluster(1);
+
+        final TestTelemetryPlugin plugin = internalCluster().getInstance(PluginsService.class, dataNode)
+            .filterPlugins(TestTelemetryPlugin.class)
+            .findFirst()
+            .orElseThrow();
+        plugin.resetMeter();
+
+        assertAcked(
+            prepareCreate(
+                "test",
+                Settings.builder()
+                    .put("index.refresh_interval", "-1")
+                    .put("index.analysis.analyzer.test_analyzer.type", "custom")
+                    .put("index.analysis.analyzer.test_analyzer.tokenizer", "standard")
+                    .putList("index.analysis.analyzer.test_analyzer.filter", "test_token_filter")
+            ).setMapping(Map.of("properties", Map.of("test_field", Map.of("type", "text", "analyzer", "test_analyzer")))).get()
+        );
+
+        String docId = randomUUID();
+        // successful index (with version)
+        client(dataNode).index(
+            new IndexRequest("test").id(docId)
+                .version(10)
+                .versionType(randomFrom(VersionType.EXTERNAL, VersionType.EXTERNAL_GTE))
+                .source(Map.of())
+        ).actionGet();
+        // if_primary_term conflict
+        {
+            var e = expectThrows(
+                VersionConflictEngineException.class,
+                () -> client(dataNode).index(new IndexRequest("test").id(docId).source(Map.of()).setIfSeqNo(0).setIfPrimaryTerm(2))
+                    .actionGet()
+            );
+            assertThat(e.getMessage(), containsString("version conflict"));
+            assertThat(e.status(), is(RestStatus.CONFLICT));
+        }
+        // if_seq_no conflict
+        {
+            var e = expectThrows(
+                VersionConflictEngineException.class,
+                () -> client(dataNode).index(new IndexRequest("test").id(docId).source(Map.of()).setIfSeqNo(1).setIfPrimaryTerm(1))
```

Comment:
Is this because the seqno is 0 for the document?

Reply:
Yes.